test/e2e/windows/eviction.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package windows
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubernetes/test/e2e/feature"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+var _ = sigDescribe(feature.Windows, "Eviction", framework.WithSerial(), framework.WithSlow(), (func() {
+	ginkgo.BeforeEach(func() {
+		e2eskipper.SkipUnlessNodeOSDistroIs("windows")
+	})
+
+	f := framework.NewDefaultFramework("eviction-test-windows")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	ginkgo.It("hhould evict a pod when a node experiences memory pressure", func(ctx context.Context) {
+		framework.Logf("Looking for a Windows node with memory-pressure eviction enabled")
+		selector := labels.Set{"kubernetes.io/os": "windows"}.AsSelector()
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(ctx, metav1.ListOptions{
+			LabelSelector: selector.String(),
+		})
+		framework.ExpectNoError(err)
+
+		var node *v1.Node
+		var nodeMem nodeMemory
+		for _, n := range nodeList.Items {
+			nm := getNodeMemory(ctx, f, n)
+			if nm.hardEviction.Value() != 0 {
+				framework.Logf("Using node %s", n.Name)
+				node = &n
+				nodeMem = nm
+			}
+		}
+
+		if node == nil {
+			e2eskipper.Skipf("No Windows nodes with hard memory-pressure eviction found")
+		}
+
+		ginkgo.By("Scheduling a pod that requests and consumse 500Mi of Memory")
+
```

Comment:
Few thoughts on this test case.

Are we protecting ourselves if we ever increase memory for the node size? The hard coded 500Mi?

Is it possible to read the amount of the memory of the node and calculate the MB for the pod based on that?

I don't understand why scheduling two pods of 500 Mi would cause memory pressure here? Do you need to set evictionHard limits?

Reply:
The first pod in the test will consume 500Mi of memory, the second pod will consume the rest of the node's memory in 100Mi chunks using https://learn.microsoft.com/en-us/sysinternals/downloads/testlimit which is in the Windows `resource-consumer` test image.

Let me try and get the test passing as is in CI (I was using a much smaller node when testing locally) then I'll update the logic to be smarter here.

Reply:
I updated this a little and added some more comments - please take a look!